apps/src/templates/instructions/TeacherFeedback.jsx

```diff
@@ -107,6 +107,24 @@ class TeacherFeedback extends Component {
     const {user, serverLevelId, teacher} = this.props;
     const {studentId} = this.state;
 
+    window.addEventListener('beforeunload', event => {
+      const latestFeedback =
+        this.state.latestFeedback.length > 0
+          ? this.state.latestFeedback[0]
+          : null;
+      const feedbackUnchanged =
+        (latestFeedback &&
+          (this.state.comment === latestFeedback.comment &&
+            this.state.performance === latestFeedback.performance)) ||
+        (!latestFeedback &&
+          (this.state.comment.length === 0 && this.state.performance === null));
+
+      if (!feedbackUnchanged) {
+        event.preventDefault();
+        event.returnValue = 'Are you sure? Your feedback may not be saved.';
```

Comment:
please move this string to [common/en_us.json](https://github.com/code-dot-org/code-dot-org/blob/98901122be9cc3465461da387eb049ff611ad3c8/apps/i18n/common/en_us.json) so it can be translated. i also wonder if the string should say "Are you sure? Your feedback _will not_ be saved" (instead of "may not"). if the teacher navigates away from the page at this time, we definitely won't save the feedback, right? or is there a chance that the feedback has/will get saved?

Reply:
Good suggestions; note that in _most_ of our supported browsers this string is entirely ignored and replaced by a message enforced by the browser. :grin: